src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -2874,49 +2874,129 @@ public override IOperation VisitCoalesceAssignment(ICoalesceAssignmentOperation
 
             int resultCaptureId = isStatement ? -1 : captureIdForResult ?? GetNextCaptureId(resultCaptureRegion);
 
-            ConditionalBranch(MakeIsNullOperation(valueCapture),
-                jumpIfTrue: true,
-                whenNull);
+            if (operation.Target?.Type.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T &&
+                ((INamedTypeSymbol)operation.Target.Type).TypeArguments[0].Equals(operation.Type))
+            {
+                nullableValueTypeReturn();
+            }
+            else
+            {
+                standardReturn();
+            }
 
-            if (!isStatement)
+            PopStackFrame(frame);
+
+            LeaveRegionsUpTo(resultCaptureRegion);
+            AppendNewBlock(afterCoalesce);
+
+            return isStatement ? null : GetCaptureReference(resultCaptureId, operation);
+
+            void nullableValueTypeReturn()
             {
-                _currentBasicBlock = null;
+                int intermediateResult = -1;
+                EvalStackFrame intermediateFrame = null;
 
-                AddStatement(new FlowCaptureOperation(resultCaptureId, operation.Syntax, OperationCloner.CloneOperation(valueCapture)));
-            }
+                if (!isStatement)
+                {
+                    intermediateFrame = PushStackFrame();
+                    SpillEvalStack();
```

Comment:
>SpillEvalStack [](start = 20, length = 14)

not related to this PR: I feel like I forgot what most of those utility methods do since last time I reviewed some CFG work. It may be worth investing in some documentation for CFG so that we can reload some context in future reviews.